### PR TITLE
[ROCm][Bugfix] Sanitize AITER paged-MQA logits before sparse top-k for DeepSeek-V4

### PR DESCRIPTION
**Pull request by shen-shanshan**
## Purpose

Fix: https://github.com/Fangzhou-Ai/vllm/issues/13

### Summary

This change prevents a GPU memory access fault when serving DeepSeek V4 with ROCm sparse attention on gfx950 under concurrent mixed prefill/decode workloads.

The ROCm AITER `deepgemm_fp8_paged_mqa_logits` kernel can intermittently return `NaN` or infinite values within the valid logits range. The generic histogram top-k kernel assumes finite, orderable inputs. Non-finite logits can leave some of its shared-memory output slots unwritten, after which uninitialized values are copied to the global top-k index buffer. Downstream sparse attention treats large non-negative values as valid token indices and uses them for block-table lookup, eventually causing an out-of-bounds GPU access.

This PR sanitizes the AITER decode logits at the ROCm call site before invoking the generic top-k kernel. It does not modify `sampler.cu` or any NVIDIA path.

### Root Cause Analysis

Runtime instrumentation captured the following failure sequence:

1. The valid range of an affected AITER paged-MQA output contained `NaN` and infinite logits.
2. `top_k_per_row_decode` then returned uninitialized values, including large positive indices and values below the `-1` sentinel.
3. The row bounds were valid, but the positive indices were far beyond their corresponding sequence lengths.
4. Sparse decode used those indices to address the block table and KV cache, causing the GPU memory fault.

These results isolate the first non-finite values to the gfx950 AITER paged-MQA output. The exact arithmetic or scheduling defect inside AITER remains to be fixed upstream; this PR enforces the vLLM-side input contract for the shared top-k consumer.

### Main Changes

Immediately after `rocm_fp8_paged_mqa_logits` and before `top_k_per_row_decode`, normalize non-finite values in place:

- `NaN` and `-Inf` become the dtype's finite minimum;
- `+Inf` becomes the dtype's finite maximum.

This preserves the ordering direction of infinities while ensuring every input to histogram top-k is finite and comparable. The top-k operator still uses `seq_lens` to restrict each row's valid range, so converting padding from `-Inf` to the finite minimum does not make padding selectable.

Keeping the workaround in `rocm_aiter_sparse_attn_indexer` limits it to the affected ROCm AITER decode path and avoids changing the stable cross-platform sampler kernel.

### Functional Test

As mentioned in https://github.com/Fangzhou-Ai/vllm/issues/13, launch the server:

```bash
vllm serve \
--model /shared/models/deepseek-ai/DeepSeek-V4-Pro \
--port 8333 \
--tensor-parallel-size 8 \
--data-parallel-size 1 \
--async-scheduling \
--no-enable-prefix-caching \
--distributed-executor-backend mp \
--gpu-memory-utilization 0.8 \
--kv-cache-dtype fp8 \
--trust-remote-code \
--tokenizer-mode deepseek_v4 \
--reasoning-parser deepseek_v4 \
--moe-backend aiter \
--compilation-config '{"mode":3,"cudagraph_mode":"FULL_AND_PIECEWISE"}' \
--max-model-len 9472
```

Then send requests with 64 conc (the script could be found in https://github.com/Fangzhou-Ai/vllm/issues/13):

```bash
docker run --rm \
--network=host \
--volume "$PWD:/repro" \
--volume /mnt/dcgpuval:/shared \
--entrypoint python \
vllm/vllm-openai-rocm:nightly \
/repro/repro_high_conc_gsm8k.py \
--prompts /repro/gsm8k_20shot_prompts_64.jsonl \
--base-url http://127.0.0.1:8333 \
--model /shared/models/deepseek-ai/DeepSeek-V4-Pro \
--num-prompts 64 \
--concurrency 64 \
--max-tokens 16 \
--waves 3 \
--timeout 300 \
--output /repro/report.json
```

### Acc Test

Full GSM8K evaluation on this change (1,319 examples).

### CI Test

```bash
pytest -sv tests/kernels/test_top_k_per_row.py::test_rocm_decode_topk_nonfinite_logits_return_bounded_indices
```

### ✅ Functional Test

**Before the fix:**

- The concurrent serving workload failed with explicit `max_model_len` values of 8192, 9472, 9473, and 16384.
- The failure reproduced with `FULL_AND_PIECEWISE`, `PIECEWISE`, and `NONE` graph modes.
- Instrumentation observed non-finite AITER logits followed by out-of-range top-k indices and a GPU memory access fault.

Part of the logs are shown below:

```bash
# Server
Memory access fault by GPU node-3 (Agent handle: 0x14b82720) on address 0x6fffe4921000. Reason: Unknown.
GPU coredump: execvp failed: No such file or directory
Failed to write program header to pipe: Broken pipe
GPU coredump: handler exited with error (status: 1)
GPU core dump failed

# Client
{"wave": 1, "requested": 64, "completed": 0, "failed": 64, "health_status": 503, "elapsed_s": 27.795312408998143}
```

**After the fix:**

- Instrumentation still observed non-finite values in the raw AITER output.
- The normalized logits contained no non-finite values.
- No top-k index was below `-1` or outside its row's sequence length.
- Three concurrent end-to-end runs completed without a GPU memory access fault.

Part of the logs are shown below:

```bash
# Server
(APIServer pid=64899) INFO:     127.0.0.1:45782 - "POST /v1/chat/completions HTTP/1.1" 200 OK
(APIServer pid=64899) INFO:     127.0.0.1:45794 - "POST /v1/chat/completions HTTP/1.1" 200 OK
(APIServer pid=64899) INFO:     127.0.0.1:45806 - "POST /v1/chat/completions HTTP/1.1" 200 OK
...

# Client
{"wave": 1, "requested": 64, "completed": 64, "failed": 0, "health_status": 200, "elapsed_s": 14.222921347012743}
{"wave": 2, "requested": 64, "completed": 64, "failed": 0, "health_status": 200, "elapsed_s": 14.297744342009537}
{"wave": 3, "requested": 64, "completed": 64, "failed": 0, "health_status": 200, "elapsed_s": 14.362902717024554}
```

### ✅ Acc Test

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9530|±  |0.0058|
|     |       |strict-match    |     5|exact_match|↑  |0.9538|±  |0.0058|

Both results exceed the required 94% threshold.

### ✅ CI Test

```
test_rocm_decode_topk_nonfinite_logits_return_bounded_indices PASSED
================ 1 passed, 14 warnings in 1.20s ================ccccc
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by Rohan138** [comment]
FYI @Fangzhou-Ai @shen-shanshan I ran into a related NaN/inf bug for sparse MLA with GLM-5.1-FP8, https://github.com/ROCm/aiter/commit/a5bd6d91801e27282017ae194dd1a09ab04b2bd6 in aiter 0.1.16.post5 fixed it for us. Wonder if rerunning fixes the NaNs at the source for you as well, I was going to introduce a similar inf/nan patch in `sampler.cu` as a workaround.

**Comment by Fangzhou-Ai** [comment]
> FYI @Fangzhou-Ai @shen-shanshan I ran into a related NaN/inf bug for sparse MLA with GLM-5.1-FP8, [ROCm/aiter@a5bd6d9](https://github.com/ROCm/aiter/commit/a5bd6d91801e27282017ae194dd1a09ab04b2bd6) in aiter 0.1.16.post5 fixed it for us. Wonder if rerunning fixes the NaNs at the source for you as well, I was going to introduce a similar inf/nan patch in `sampler.cu` as a workaround.

Thank you for your info @Rohan138 ! Please check if this PR fix @shen-shanshan

**Comment by shen-shanshan** [comment]
> > FYI @Fangzhou-Ai @shen-shanshan I ran into a related NaN/inf bug for sparse MLA with GLM-5.1-FP8, [ROCm/aiter@a5bd6d9](https://github.com/ROCm/aiter/commit/a5bd6d91801e27282017ae194dd1a09ab04b2bd6) in aiter 0.1.16.post5 fixed it for us. Wonder if rerunning fixes the NaNs at the source for you as well, I was going to introduce a similar inf/nan patch in `sampler.cu` as a workaround.
> 
> Thank you for your info @Rohan138 ! Please check if this PR fix @shen-shanshan

That PR doesn't fix this problem.

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 1]
```diff
@@ -623,6 +623,32 @@ def _topk_indices_torch(
     return padded
 
 
+def _rocm_top_k_per_row_decode(
+    logits: torch.Tensor,
+    next_n: int,
+    seq_lens: torch.Tensor,
+    topk_indices: torch.Tensor,
+) -> None:
+    finite_min = torch.finfo(logits.dtype).min
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: From performance perspective, `torch.finfo(logits.dtype).min` is costly.

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 2]
```diff
@@ -623,6 +623,32 @@ def _topk_indices_torch(
...
+    finite_max = torch.finfo(logits.dtype).max
+    logits.nan_to_num_(
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: For references purpose only as we are prioritizing acc. How much cost is incurred by this op?

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py (hunk 1 above)
Comment (shen-shanshan) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: > From performance perspective, `torch.finfo(logits.dtype).min` is costly.

After verification, this line is not needed.

**Comment by shen-shanshan** [comment]
**Performance Comparison (under 64 conc):**

| Metric | Before this PR | After this PR | Comparison |
| :----- | :------------- | :------------ | :--------- |
| Output token throughput (tok/s) | 822.26 | 819.69 | -0.31% ↓ |
| Mean TTFT (ms) | 2,111.71 | 2,119.92 | +0.39% ↑ |
| Mean TPOT (ms) | 73.98 | 74.17 | +0.26% ↑ |

@tjtanaa The performance becomes slightly lower. Do you think the change is acceptable in order to fix this functional error (short-term solution for quick fix)? And in long-term, I suppose it's better to fix the `deepgemm_fp8_paged_mqa_logits()` in AITER side.

**Comment by tjtanaa** [comment]
> **Performance Comparison (under 64 conc):**
> 
> Metric	Before this PR	After this PR	Comparison
> Output token throughput (tok/s)	822.26	819.69	-0.31% ↓
> Mean TTFT (ms)	2,111.71	2,119.92	+0.39% ↑
> Mean TPOT (ms)	73.98	74.17	+0.26% ↑
> @tjtanaa The performance becomes slightly lower. Do you think the change is acceptable in order to fix this functional error (short-term solution for quick fix)? And in long-term, I suppose it's better to fix the `deepgemm_fp8_paged_mqa_logits()` in AITER side.

Thanks for getting the numbers. I think this is acceptable tradeoff for fixing the functional error.

**Review by tjtanaa — APPROVED** [review]

**Comment by shen-shanshan** [comment]
https://github.com/vllm-project/vllm/pull/50060 has fixed the CI error.

**Comment by shen-shanshan** [comment]
@tjtanaa Full CI passed, this can be merged.
